### PR TITLE
Add SortedQueue class

### LINKED ISSUE #46: Add PriorityQueue class to main
- [x] include tests
- [x] include documentation

### PR DESCRIPTION
**Pull request by clorton**
Fixes #46

### CODE REVIEW

**Comment by clorton** [comment]
**_PR #60 should be merged first!_**

**Comment by jonathanhhb** [comment]
2 quick questions from quick scan:

1) What's up with the names of the functions in the API? They lack intuition to me. Are they borrowed from heapq? 

2) The tests look like they cover basic functionality -- is there a stress/perf test we can add?

**Comment by clorton** [comment]
Re. no. 1 - do you mean `peeki()`, `peekv()` etc. or `_siftup()` and `_siftdown()`?
The former are my functions to get only the index or value (because creating a tuple of (i, v) is expensive).
The latter are from the internal Python implementation of `heapq`.

**Comment by jonathanhhb** [comment]
I think I would expect to be able to create an empty queue and push values.

Also:

>>> from src.laser_core.priorityqueue import PriorityQueue
>>> pq = PriorityQueue(capacity=1, values=[44])
>>> pq.push( 44 )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jbloedow/PLAY/LASER/laser-chris-nnmm/src/laser_core/priorityqueue.py", line 43, in push
    _siftdown(self.indices, self.values, 0, self.size)
  File "/var/opt/idm/venv_laser/lib/python3.10/site-packages/numba/core/dispatcher.py", line 658, in _explain_matching_error
    raise TypeError(msg)
TypeError: No matching definition for argument type(s) array(uint32, 1d, C), reflected list(int64)<iv=None>, int64, int64

This seems surprising. Though also pushing a value without an associated priority also seems surprising. 

This is more like what I'd expect:

image

**Comment by clorton** [comment]
This is a documentation issue. You should push the _index_ of the item you want in the queue. `value[index]` get the "priority". This is because generally we are using values _already stored in the model_ for our priorities and I don't want to copy them.

**Comment by clorton** [comment]
The error you got above doesn't seem useful. I should look into that (although, I was limiting the sanity checking in `push()`, `peek()`, and `pop()` for performance reasons.

The "priority is in the `values`" parameter is very intentional for LASER (as mentioned above). Python already has a perfectly good heap implementation for the API you got from ChatGPT.

File: tests/test_sortedqueue.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        """Test peeking at the top element of the sorted queue."""
+        self.pq.push(0)
+        self.pq.push(1)
+        self.pq.push(2)
+        self.pq.push(3)
+        self.pq.push(4)
+        self.pq.push(5)
+        self.pq.push(6)
+
+        assert self.pq.peekiv() == (3, 26)
```
Comment (jonathanhhb) [inline tests/test_sortedqueue.py:60]: When testing a peek, would it be better to do two in a row to prove that the peek is just a peek and not a pop?

File: tests/test_sortedqueue.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            _ = self.pq.peekiv()
+
+    def test_peek_random(self):
+        """Test peeking at the top element of the sorted queue with random values."""
+        values = np.random.randint(0, 100, 1024, dtype=np.int32)
+        self.pq = SortedQueue(len(values), values)
+        for i in range(len(values)):
+            self.pq.push(i)
+        minimum = values.min()
+        assert self.pq.peekv() == minimum
```
Comment (jonathanhhb) [inline tests/test_sortedqueue.py:74]: I guess same peek comment here.

File: tests/test_sortedqueue.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        """Test the timing of the push method."""
+        import timeit
+
+        np.random.seed(20240701)
+        count = 1 << 20
+        values = np.random.randint(0, 100, count, dtype=np.int32)
+        self.pq = SortedQueue(len(values), values)
+        elapsed = timeit.timeit("for i in range(len(values)): self.pq.push(i)", globals={"values": values, "self": self}, number=1)
+        messages.append(
+            f"SortedQueue.push() timing: {elapsed:0.4f} seconds for {count:9,} elements = {int(round(count / elapsed)):11,} elements/second"
```
Comment (jonathanhhb) [inline tests/test_sortedqueue.py:104]: I don't think I saw this output in my console when I ran that test.

File: tests/test_sortedqueue.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        """Test peeking at the top index of the sorted queue."""
+        self.pq.push(0)
+        self.pq.push(1)
+        self.pq.push(2)
+        self.pq.push(3)
+        self.pq.push(4)
+        self.pq.push(5)
+        self.pq.push(6)
+
+        assert self.pq.peeki() == 3
```
Comment (jonathanhhb) [inline tests/test_sortedqueue.py:134]: Same comment about re-peeking.

File: src/laser_core/sortedqueue.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+import numba as nb
+import numpy as np
+
+
+class SortedQueue:
+    """
+    A sorted (priority) queue implemented using NumPy arrays and sped-up with Numba.
+    Using the algorithm from the Python heapq module.
+    __init__ with an existing array of sorting values
```
Comment (jonathanhhb) [inline src/laser_core/sortedqueue.py:13]: typo? *sortable?

File: src/laser_core/sortedqueue.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    __push__ with an index into sorting values
+    __pop__ returns the index of the lowest sorting value and its value
+    """
+
+    # https://github.com/python/cpython/blob/5592399313c963c110280a7c98de974889e1d353/Modules/_heapqmodule.c
+    # https://github.com/python/cpython/blob/5592399313c963c110280a7c98de974889e1d353/Lib/heapq.py
+
+    def __init__(self, capacity: int, values: np.ndarray):
+        self.indices = np.zeros(capacity, dtype=np.uint32)
```
Comment (jonathanhhb) [inline src/laser_core/sortedqueue.py:22]: Given the particular nature of this class, I'd like more of a docstring on the constructor here.

File: tests/test_sortedqueue.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        """Test popping the top index of the sorted queue."""
+        self.pq.push(0)
+        self.pq.push(1)
+        self.pq.push(2)
+        self.pq.push(3)
+        self.pq.push(4)
+        self.pq.push(5)
+        self.pq.push(6)
+
+        assert self.pq.popi() == 3
```
Comment (jonathanhhb) [inline tests/test_sortedqueue.py:153]: Seems like we should do a test of what's there after the first pop as well.

**Review by jonathanhhb — CHANGES_REQUESTED** [review]

**Comment by jonathanhhb** [comment]
Is there any reason we don't implicitly push the sequence of integers starting at 0 (as the corresponding indices) in the constructor to match the values array?

File: tests/test_sortedqueue.py (hunk 3 above)
Comment (clorton) [reply tests/test_sortedqueue.py:115]: It's a little bit of work to make sure the test harness doesn't swallow up stdout. Should be fixed now.

**Comment by clorton** [comment]
We don't push all the indices because we don't necessarily track _all_ the agents. For example, there's no need to put agents who will die after the simulation is done into the queue. That would just increase `push()` and `pop()` overhead.

**Comment by jonathanhhb** [comment]
I also just noticed that pop() doesn't return anything. That seems surprising. 

And maybe I'm just being slow still, but I haven't intuited when one would use popi, popv, popiv or pop.

**Comment by clorton** [comment]
Generally, we just want the index of the next item on the queue, so `popi()`. The others are there for symmetry.

**Comment by jonathanhhb** [comment]
> Generally, we just want the index of the next item on the queue, so `popi()`. The others are there for symmetry.
> 
> We would tend to use `popi()` after using `peekv()` to see if the _value_ is relevant, e.g. date of death == now. Then `popi()` gets us the index to use to update state.

Should pop() be a private method?

**Comment by clorton** [comment]
> > Generally, we just want the index of the next item on the queue, so `popi()`. The others are there for symmetry.
> > We would tend to use `popi()` after using `peekv()` to see if the _value_ is relevant, e.g. date of death == now. Then `popi()` gets us the index to use to update state.
> 
> Should pop() be a private method?

Done.

**Comment by jonathanhhb** [comment]
I'm probably done with my due diligence on this PR at this point. :)

**Pull request merged**